compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs

```diff
@@ -0,0 +1,254 @@
+    forM_ existingChoices $ \choice -> do
+        withContext (ContextTemplate (present module_) (present template) (TPChoice (present choice))) $ do
+            let returnTypesMatch = foldU alphaType (fmap chcReturnType choice)
+            unless returnTypesMatch $
+                throwWithContext (EUpgradeChoiceChangedReturnType (NM.name (present choice)))
+
+            throwIfDifferent "controllers" (extractFuncFromFuncThisArg . chcControllers <$> choice) $
+                EUpgradeChoiceChangedControllers $ NM.name $ present choice
+
+            let observersErr = EUpgradeChoiceChangedObservers $ NM.name $ present choice
+            case fmap (mapENilToNothing . chcObservers) choice of
+               Upgrading { past = Nothing, present = Nothing } -> do
+                   pure ()
+               Upgrading { past = Just past, present = Just present } -> do
+                   throwIfDifferent "observers" (extractFuncFromFuncThisArg <$> Upgrading past present) observersErr
+               _ -> do
+                   throwWithContext observersErr
+
+            let authorizersErr = EUpgradeChoiceChangedAuthorizers $ NM.name $ present choice
+            case fmap (mapENilToNothing . chcAuthorizers) choice of
+               Upgrading { past = Nothing, present = Nothing } -> pure ()
+               Upgrading { past = Just past, present = Just present } ->
+                   throwIfDifferent "authorizers" (extractFuncFromFuncThisArg <$> Upgrading past present) authorizersErr
+               _ -> throwWithContext authorizersErr
+        pure choice
+
+    -- This check assumes that we encode signatories etc. on a template as
+    -- $<uniquename> this, where $<uniquename> is a function that contains the
+    -- actual definition. We resolve this function and check that it is
+    -- identical.
+    withContext (ContextTemplate (present module_) (present template) TPPrecondition) $
+        throwIfDifferent "precondition" (extractFuncFromCaseFuncThis . tplPrecondition <$> template) $
+            EUpgradeTemplateChangedPrecondition $ NM.name $ present template
+    withContext (ContextTemplate (present module_) (present template) TPSignatories) $
+        throwIfDifferent "signatories" (extractFuncFromFuncThis . tplSignatories <$> template) $
+            EUpgradeTemplateChangedSignatories $ NM.name $ present template
+    withContext (ContextTemplate (present module_) (present template) TPObservers) $
+        throwIfDifferent "observers" (extractFuncFromFuncThis . tplObservers <$> template) $
+            EUpgradeTemplateChangedObservers $ NM.name $ present template
+    withContext (ContextTemplate (present module_) (present template) TPAgreement) $
+        throwIfDifferent "agreement" (extractFuncFromFuncThis . tplAgreement <$> template) $
+            EUpgradeTemplateChangedAgreement $ NM.name $ present template
+    -- TODO: Check that return type of a choice is compatible
```

Comment:
I see you're currently just checking the types are the same, do we intend to support upgrading the return type?

Reply:
Yes, I think so, but I prefer to push that to another PR because a full check for type compatibility is a bit harder than structs.

Reply:
Issue for PR to come: https://github.com/digital-asset/daml/issues/17611